lib/logout.js

```diff
@@ -0,0 +1,350 @@
+                serviceProviderLogoutURL: session.serviceProviderLogoutURL
+              },
+              relayState: req.query.RelayState || (req.body && req.body.RelayState)
+            };
+
+            options.store.save(req, spData, function (err, transactionId) {
+              if (err) { return next(err); }
+
+              // We remove the session from the LogoutRequest Originator.
+              // This session is already saved in the store.
+              // We should not send a LogoutRequest to that session
+              // Only a LogoutResponse when there are no other session participants active
+              options.sessionParticipants.remove(requestData.issuer, function (err) {
+                if (err) { return next(err); }
+
+                prepareAndSendLogoutRequest(options.sessionParticipants, transactionId, req, res, next);
+              });
+            });
+          });
+        });
+
+      // Logout flow in progress, incoming SAMLResponse from SP. (Could be SP initiated or IdP initiated)
+      } else if (req.query.SAMLResponse || req.body.SAMLResponse) {
+        function processLogoutResponse(logoutResponse, transactionId) {
+          // LogoutResponse was OK, we remove the session participant from the IdP
+          options.sessionParticipants.remove(logoutResponse.issuer, function (err) {
+            if(err) return next(err);
+
+            // Continue with next session if any
+            prepareAndSendLogoutRequest(options.sessionParticipants, transactionId, req, res, next);
+          });
+        }
+
+        // Verify that the state sent to the SP matches the one returned
+        var h = req.query.RelayState || (req.body && req.body.RelayState);
+        options.store.load(req, h, {destroy: true}, function (err, state) {
+          if (err) { return next(err); }
+          if (!state) { return next(state); }
+
+          // If there are sessions left, keep sending LogoutRequest to Session Participants. If not finish
+          validateSamlResponse(req, options.sessionParticipants, function (err, logoutResponse) {
+            if (err) { return next(err); }
+
+            var transactionId = state.transactionId;
+            if (logoutResponse.status === STATUS.SUCCESS) {
+              return processLogoutResponse(logoutResponse, transactionId);
+            }
+
+            // Mark global status as partial logout if a logout does not succeed
+            options.store.load(req, transactionId, function (err, state) {
+              state.global_status = 'failed';
+              options.store.update(req, transactionId, state, function (err) {
+                if (err) { return next(err); }
+
+                processLogoutResponse(logoutResponse, transactionId);
+              });
+            });
+          });
+        });
+
+      // IdP initated - Start flow - In this case we will show a 200 when complete
+      } else {
+        options.store.save(req, {}, function (err, transactionId) {
```

Comment:
> What if the transaction-storage stuff is managed by the prepareAndSendLogoutRequest itself? 
I think that it will be messy and not clear

Let me explain this:

We store two different things in this store.
1. Global transaction: It is the state of the entire SLO transaction identified by transaction id. We store the global status, and if SP initiated, the information of the SP 
2. SP/IDP current transaction: We store under the RelayState key, the transaction ID (e..g `state[RelayState] = {transactionID: txid }`)

Number 2 is handled by `prepareAndSendLogoutRequest` as it is handled in a per request basis.

Number 1 is not that easy to handle it there.